internal/saver/saver.go

```diff
@@ -0,0 +1,83 @@
+package saver
+
+import (
+	"errors"
+	"time"
+
+	"github.com/ozoncp/ocp-remind-api/internal/flusher"
+	"github.com/ozoncp/ocp-remind-api/internal/models"
+)
+
+type Saver interface {
+	Save(remind models.Remind) error
+	Init()
+	Close()
+}
+
+type remindSaver struct {
+	flusher   flusher.Flusher
+	ticker    *time.Ticker
+	reminds   []models.Remind
+	saveChan  chan models.Remind
+	closeChan chan struct{}
+}
+
+func (saver *remindSaver) Init() {
+	go func() {
+		defer saver.ticker.Stop()
+
+		for {
+			select {
+			case remind := <-saver.saveChan:
+				saver.reminds = append(saver.reminds, remind)
+			case <-saver.ticker.C:
+				if len(saver.reminds) != 0 {
+					saver.reminds = saver.flusher.Flush(saver.reminds)
+				}
+			case <-saver.closeChan:
+				if len(saver.reminds) != 0 {
+					saver.flusher.Flush(saver.reminds)
+				}
+
+				close(saver.saveChan)
+				close(saver.closeChan)
+
+				return
+			}
+		}
+	}()
+}
+
+var ErrCapacity = errors.New("there is no avaible capacity")
+
+func (saver remindSaver) Save(remind models.Remind) error {
+	if len(saver.reminds) < cap(saver.reminds) {
+		saver.saveChan <- remind
+
+		return nil
+	}
+
+	return ErrCapacity
+}
+
+func (saver remindSaver) Close() {
+	saver.closeChan <- struct{}{}
+}
```

Comment:
Каналы надо закрывать на `Close`, т.к. тут работает правило «закрывает писатель/производитель» (короче, producer closes). [Чтиво](https://go101.org/article/channel-closing.html). Закрытие канала ведёт к тому, что из него будет вычитывание читателем, поэтому верхние 2 `close` и не нужны

```go
func (saver remindSaver) Close() {
	close(saver.closeC)
	close(saver.bufC)
}
```

Пример с вычиткой:

```go
var closeC, now, wg = make(chan struct{}), time.Now(), sync.WaitGroup{}
wg.Add(1)
go func() { defer wg.Done(); <-closeC; println("i'm closed!"); fmt.Println(time.Since(now)) }()
time.Sleep(time.Second * 1)
close(closeC)
wg.Wait()
```